app/client/rest/tracking.ts

```diff
@@ -93,40 +101,118 @@ export default class ClientTracking {
+        if (!groupData) {
+            return 100;
+        }
+
+        const urlData = Object.entries(groupData.urls);
+        const sumLatency = urlData.reduce((result, url) => (result + (url[1].metrics?.latency ?? 0)), 0);
+        return sumLatency / urlData.length;
+    }
+
+    categorizeRequests(groupLabel: RequestGroupLabel): CategorizedRequests {
+        const parallel: ClientResponseMetrics[] = [];
+        const sequential: ClientResponseMetrics[] = [];
+        const groupData = this.requestGroups.get(groupLabel);
+        if (!groupData) {
+            return {parallel, sequential};
+        }
+
+        // Sort requests by start time
+        const requestsMetrics = Object.entries(groupData.urls).map((e) => e[1].metrics).filter((m) => m != null);
+        requestsMetrics.sort((a, b) => a.startTime - b.startTime);
+
+        let lastEndTime = 0;
+
+        for (const metrics of requestsMetrics) {
+            if (metrics.startTime < lastEndTime) {
+                // Overlapping request -> Parallel
+                parallel.push(metrics);
+            } else {
+                // Non-overlapping request -> Sequential
+                sequential.push(metrics);
+            }
+
+            // Update the last end time
+            lastEndTime = Math.max(lastEndTime, metrics.endTime);
+        }
+
+        return {parallel, sequential};
+    }
+
+    calculateAverageSpeedWithCategories = (
+        categorizedRequests: CategorizedRequests,
+        elapsedTimeInSeconds: number, // Observed total elapsed time in seconds
+    ): { averageSpeedMbps: number; effectiveLatency: number } => {
+        const {parallel, sequential} = categorizedRequests;
+
+        // Step 1: Calculate total data size in bits
+        const totalDataBits = [...parallel, ...sequential].reduce((sum, req) => sum + (req.compressedSize * 8), 0);
+
+        // Step 2: Calculate effective latency
+        const parallelLatencies = parallel.map((req) => req.latency);
+        const sequentialLatencies = sequential.map((req) => req.latency);
+
+        const effectiveParallelLatency = Math.max(...parallelLatencies, 0);
+        const effectiveSequentialLatency = sequentialLatencies.reduce((sum, latency) => sum + latency, 0);
+
+        const effectiveLatency = effectiveParallelLatency + effectiveSequentialLatency;
+
+        // Step 3: Calculate data transfer time
+        const dataTransferTime = elapsedTimeInSeconds - (effectiveLatency / 1000);
+
+        // Handle edge case: if data transfer time is zero or negative
+        if (dataTransferTime <= 0) {
+            return {averageSpeedMbps: 0, effectiveLatency: 0};
+        }
+
+        // Step 4: Calculate average speed
+        const averageSpeedBps = totalDataBits / dataTransferTime; // Speed in bps
+        const averageSpeedMbps = averageSpeedBps / 1_000_000; // Convert to Mbps
+
+        return {
+            averageSpeedMbps,
+            effectiveLatency,
+        };
+    };
```

Comment:
I've made a significant proposed improvement based on data observation.  I will try to explain as much as possible in this PR: https://github.com/mattermost/mattermost-mobile/pull/8439.